elements/lisk-transactions/src/14_unlock_transaction.ts

```diff
@@ -0,0 +1,306 @@
+						unvoteHeight: unlock.unvoteHeight,
+					};
+				}),
+			};
+		} else {
+			// tslint:disable-next-line no-object-literal-type-assertion
+			this.asset = { unlockingObjects: [] };
+		}
+	}
+
+	public assetToJSON(): object {
+		return {
+			unlockingObjects: this.asset.unlockingObjects.map(unlock => ({
+				delegateAddress: unlock.delegateAddress,
+				amount: unlock.amount.toString(),
+				unvoteHeight: unlock.unvoteHeight,
+			})),
+		};
+	}
+
+	protected assetToBytes(): Buffer {
+		const bufferArray = [];
+		for (const unlock of this.asset.unlockingObjects) {
+			const addressBuffer = intToBuffer(
+				unlock.delegateAddress.slice(0, -1),
+				SIZE_UINT64,
+			);
+			bufferArray.push(addressBuffer);
+			const amountBuffer = intToBuffer(
+				unlock.amount.toString(),
+				SIZE_INT64,
+				'big',
+				true,
+			);
+			bufferArray.push(amountBuffer);
+			const unvoteHeightBuffer = intToBuffer(unlock.unvoteHeight, SIZE_UINT32);
+			bufferArray.push(unvoteHeightBuffer);
+		}
+
+		return Buffer.concat(bufferArray);
+	}
+
+	public async prepare(store: StateStorePrepare): Promise<void> {
+		const addressArray = this.asset.unlockingObjects.map(unlock => ({
+			address: unlock.delegateAddress,
+		}));
+		const filterArray = [
+			{
+				address: this.senderId,
+			},
+			...addressArray,
+		];
+
+		await store.account.cache(filterArray);
+	}
+
+	protected validateAsset(): ReadonlyArray<TransactionError> {
+		const asset = this.assetToJSON();
+		const schemaErrors = validator.validate(voteAssetFormatSchema, asset);
+		const errors = convertToAssetError(
+			this.id,
+			schemaErrors,
+		) as TransactionError[];
+
+		for (const unlock of this.asset.unlockingObjects) {
+			if (unlock.amount <= BigInt(0)) {
+				errors.push(
+					new TransactionError(
+						'Amount cannot be less than zero',
+						this.id,
+						'.asset.votes.amount',
+						unlock.amount.toString(),
+					),
+				);
+			}
+
+			if (unlock.amount % TEN_UNIT !== BigInt(0)) {
+				errors.push(
+					new TransactionError(
+						'Amount should be multiple of 10 * 10^8',
+						this.id,
+						'.asset.votes.amount',
+						unlock.amount.toString(),
+					),
+				);
+			}
+		}
+
+		return errors;
+	}
+
+	protected async applyAsset(
+		store: StateStore,
+	): Promise<ReadonlyArray<TransactionError>> {
+		const errors = [];
+
+		for (const unlock of this.asset.unlockingObjects) {
+			const sender = await store.account.get(this.senderId);
+			const delegate = await store.account.get(unlock.delegateAddress);
+			if (
+				!hasWaited(sender, delegate, store.chain.lastBlockHeader.height, unlock)
+			) {
+				errors.push(
+					new TransactionError(
+						'Unlocking object has not waited locking period',
+						this.id,
+						'.asset.unlockingObjects.unvoteHeight',
+						unlock.unvoteHeight,
+					),
+				);
+			}
+			if (isPunished(sender, delegate, store.chain.lastBlockHeader.height)) {
+				errors.push(
+					new TransactionError(
+						'Delegate is currently being punished',
```

Comment:
'Unlocking is not permitted as delegate is currently being punished' something like this?